tools/Nethermind.Tools.Kute/Program.cs

```diff
@@ -44,41 +45,44 @@ static IServiceProvider BuildServiceProvider(Config config)
                 config.AuthTtl
             )
         );
-        collection.AddSingleton<IMessageProvider<JsonRpc?>>(
-            new JsonRpcMessageProvider(
-                new FileMessageProvider(config.MessagesFilePath))
-        );
+        collection.AddSingleton<IMessageProvider<string>>(new FileMessageProvider(config.MessagesFilePath));
+        collection.AddSingleton<IMessageProvider<JsonRpc?>, JsonRpcMessageProvider>();
         collection.AddSingleton<IJsonRpcMethodFilter>(
             new ComposedJsonRpcMethodFilter(config.MethodFilters.Select(pattern =>
                 new PatternJsonRpcMethodFilter(pattern)))
         );
-        if (config.DryRun)
-        {
-            collection.AddSingleton<IJsonRpcSubmitter, NullJsonRpcSubmitter>();
-        }
-        else
-        {
-            collection.AddSingleton<IJsonRpcSubmitter>(provider =>
-                new HttpJsonRpcSubmitter(
+        collection.AddSingleton<IJsonRpcSubmitter>(provider =>
+            config.DryRun
+                ? new NullJsonRpcSubmitter(provider.GetRequiredService<IAuth>())
+                : new HttpJsonRpcSubmitter(
                     provider.GetRequiredService<HttpClient>(),
                     provider.GetRequiredService<IAuth>(),
                     config.HostAddress
-                )
-            );
-        }
+                ));
+        collection.AddSingleton<IProgressReporter>(provider =>
+        {
+            if (config.ShowProgress)
+            {
+                // TODO:
+                // Terrible, terrible hack since it forces a double enumeration:
+                // - A first one to count the number of messages.
+                // - A second one to actually process each message.
+                // We can reduce the cost by not parsing each message on the first enumeration
+                // At the same time, this optimization relies on implementation details.
+                var messagesProvider = provider.GetRequiredService<IMessageProvider<string>>();
+                var totalMessages = messagesProvider.Messages.ToEnumerable().Count();
+                return new ConsoleProgressReporter(totalMessages);
```

Comment:
As the comment points out, this is terrible for performance but it's the only way to know how many messages are going to be processed, and we need that information to be able to show a meaningful progress bar (`X out of TOTAL`).

For this reason, it's suggested to **not use** `-p|--progress` in automated environments.